### PR TITLE
Basic JEI integration

### PR DESCRIPTION
**Pull request by uecasm**
Adds basic JEI integration (recipe lookup).  No recipe transfer yet but I might do that in a separate PR.

Firstly, this adds all of the cutter block types to the main Domum Ornamentum creative tab (which previously only had the cutter itself on it).  Only one block of each type is added, and each one generates with randomly selected materials each time you connect your client, to keep it interesting.
image

Secondly, this adds basic JEI recipe lookup.  JEI also shows the same random items in its ingredient list on the right, and pressing U on any valid input item (at all, including ones not used by the random visible items) will show all possible usages for that item (in all cutter slots), or you can press R on any cutter item to see all possible input blocks for that output type (in theory, although I think JEI imposes a hard cap of about 100 that it will cycle between).  The output item render updates to match whatever input blocks it is showing at the time.
image
(This will break if you change the cutter's gui texture, so remember to update both at the same time if needed.)

### CODE REVIEW

**Review by Raycoms — APPROVED** [review]

**Comment by marchermans** [comment]
There was a decision made to not include them in any creative tab. This was internally discussed and decided as the best solution

So please undo that change

**Comment by Raycoms** [comment]
> There was a decision made to not include them in any creative tab. This was internally discussed and decided as the best solution
> 
> So please undo that change

As it's only a small number and not "all" (1 of each type) I think this would be fine.

**Comment by marchermans** [comment]
> > There was a decision made to not include them in any creative tab. This was internally discussed and decided as the best solution
> > So please undo that change
> 
> As it's only a small number and not "all" (1 of each type) I think this would be fine.

I clearly remember having a discussion with you about this, and your own answer was that this is not a solution since we will always then have people who ask: Where is variant X in the creative menu. And I think we should just stick with the solution that we have right now.

**Comment by Raycoms** [comment]
> > > There was a decision made to not include them in any creative tab. This was internally discussed and decided as the best solution
> > > So please undo that change
> > 
> > 
> > As it's only a small number and not "all" (1 of each type) I think this would be fine.
> 
> I clearly remember having a discussion with you about this, and your own answer was that this is not a solution since we will always then have people who ask: Where is variant X in the creative menu. And I think we should just stick with the solution that we have right now.

Yeah, I do like the random ones though. It kinda introduces different blocks =D

**Comment by marchermans** [comment]
I like the idea and the concept but I don't like the effect that it draws with it.

But i don't like the effect that it brings with it of the people complaining about missing stuff

**Comment by Raycoms** [comment]
> I like the idea and the concept but I don't like the effect that it draws with it.
> 
> But i don't like the effect that it brings with it of the people complaining about missing stuff

That effect we have right now already unfortunately. Not sure we can reduce that in any way.

**Comment by marchermans** [comment]
> > I like the idea and the concept but I don't like the effect that it draws with it.
> > But i don't like the effect that it brings with it of the people complaining about missing stuff
> 
> That effect we have right now already unfortunately. Not sure we can reduce that in any way.

But we are now consistent. Nothing is in the creative tab
This, this just brings insconsistency, and you will get people who complain even more.

I am okey with the JEI integration.
Not with the changes to the creative menu. Especially since they where not discussed in discord before hand, or only discussed in DMs.

**Comment by uecasm** [comment]
I can't remove them from the creative tab, because then they will not show up in the right-hand JEI list.  This would still _work_, technically, but they'd be massively less discoverable, and that is not acceptable.

**Comment by marchermans** [comment]
So I thought about this.
And I agree that we need some form of JEI anker item.

But I am against having it be just one.
Make them rotate through their materials every few seconds. (Should not matter performance wise, and the caches will recycle it anyway down the road). This on one hand prevents the "missing items" deallio, and gives access to items in JEI.
Besides that I would even go one step further: Give the create tab stacks a special NBT tag that says that more variants can be crafted in the cutter, and prevent placing of those stacks with that special NBT tag, since those are just decorative item variants.

This should make it really clear that these itemstacks only function as an introduction to the Cutter and the JEI interface and not as a source of those blocks, since that is what I really do not want to be there.

**Comment by uecasm** [comment]
Within the JEI recipe GUI, the items do rotate through all their materials, as already explained.  Requesting the recipe on any single item in the sidebar will show all possible materials for that specific output item type, or you can switch to the "all recipes" view to see everything that the cutter can make.  (Conversely, you can request usages of a single material to see all recipes that will accept that material.)

It does not seem reasonable to me to try to do the same in the sidebar or creative tab itself.  Nor, as far as I am aware, is this even possible.  (At least not without giving them a special NBT tag that makes the item itself do variable rendering.  Which is possible, but is beyond the scope of this PR and I don't think it adds anything useful.)

**Comment by uecasm** [comment]
So, one thing I could do relatively easily -- if I remove the caching of the creative tab items, then every time you open the tab (including switching between tabs) it will re-randomize all the item textures.  But this will only affect the creative tab, the example items in JEI will remain fixed until you relog your client.  Is that the sort of thing you're looking for?

**Comment by uecasm** [comment]
It was already flipping ingredients before this, but this latest change randomises the order of each slot (it will be consistent within a single session, but nobody will notice) so that they will usually always be different items.  Here's a video to demonstrate:

[JEI preview](https://www.youtube.com/watch?v=TWFLa9WVyxk)

Note: I have not changed anything in the creative tabs (yet?).

**Comment by marchermans** [comment]
As I said, I would like the items in the creative tab to be **ONLY** anchor items to the JEI interface or referencable to the Stonecutter and explicitly not placeable. It should be really clear that these items are _SHOW_ items only.

**Comment by uecasm** [comment]
But why?  There's no particular harm if someone does pull an item out of the creative tab or JEI cheat mode, and even use it to build with.  They won't find the exact same item the next play session, but that doesn't really matter since they can just Pick Block on any existing placed instance if they want something they've used previously, even if they've already discarded it from their inventory.  (And as currently implemented, both sources will provide the same set of items and will be stable for the whole session, which seems the least confusing behaviour.  I haven't tested on servers; it wouldn't be entirely surprising if JEI and creative did get two different sets of items there, but it should again be stable within a session either way.)

Also, every item in both places has a tooltip that directly refers to the Architect's Cutter, which is also the first item that appears in the creative tab or JEI list.  (And if they look at recipes in JEI they'll see the other input blocks.)  So I think there's enough hinting to go make one of those for more blocks already.

**Comment by marchermans** [comment]
> But why?  There's no particular harm if someone does pull an item out of the creative tab or JEI cheat mode, and even use it to build with.  They won't find the exact same item the next play session, but that doesn't really matter since they can just Pick Block on any existing placed instance if they want something they've used previously, even if they've already discarded it from their inventory.  (And as currently implemented, both sources will provide the same set of items and will be stable for the whole session, which seems the least confusing behaviour.  I haven't tested on servers; it wouldn't be entirely surprising if JEI and creative did get two different sets of items there, but it should again be stable within a session either way.)
> 
> Also, every item in both places has a tooltip that directly refers to the Architect's Cutter, which is also the first item that appears in the creative tab or JEI list.  (And if they look at recipes in JEI they'll see the other input blocks.)  So I think there's enough hinting to go make one of those for more blocks already.

I actually made it really clear why. I don't want people wining that things are missing from the creative tab. Which will happen. So if this is to be merged together then that is what I want to be changed. Make it abundently clear to any player in the first place that those items are there for show. For reference to the recipes. And nothing else. Not for placement or for pulling items out of the tab to place. Nothing. That is what I discussed with ray when working on the creative tabs. That is what our decision was about two weeks ago. And i am not going to go away from that. If this is supposed to be merged then any pull request, at least for now, has to abide by that decision. It is simpel as that. It has nothing to do with how your solution works it is a good solution to deal with the rendering. But we actively and after a long discussion dismissed this particular option when choosing what to do. And as of now i find it way toooooo early to come back to that.

**Comment by uecasm** [comment]
The most common 1.17 question on the discord at present (other than lingering installation issues) is how to craft DO blocks.  And it's very common for mods or modpacks to include only one JEI recipe for particularly spammy blocks (e.g. facades).  So nobody is going to care that only one instance of the materials are visible, that's what everyone else does.

So no.  That is absolutely never going to happen, at least not by me.  You're welcome to do it yourself if you like, but I think it's a complete waste of time and effort.  Besides which, I don't know enough about the internals of model rendering to touch that code, so I'm not going to.

Merging this now as-is will solve a lot of direct questions players have and make things more discoverable.  It's an instant win.  If it turns out that I'm wrong and the discord is inundated with people who can't find exactly the right item they want, that's when it can be revisited and refined or redesigned.  Iterate and get feedback.  Agile, KISS.

It's up to you, of course.  But this is _my_ final word on this code.

**Comment by marchermans** [comment]
If it is final then this won't be merged. 
We discussed this when we published this not even two weeks ago. 

As maintainers we decided that there should be either none or all items in the tab. And if that is not possible it should be made really clear to players that the ones shown in the creative menu are not the only ones available. KISS or no KISS. That was and still is a project guideline that has not changed with regards to items in the creative tab.

Since you don't seem to want to apply that decision that was made not even two weeks ago to your pr here, i see no future for the pull request in tis current state.

@Raycoms last final input on this please.

**Comment by Raycoms** [comment]
Okay. Let's first talk about what we all agree.
The JEI solution is fine and displaying random example blocks there is fine.

Let's talk about what we disagree on :

Having example blocks in the creative tab without being clear that those are only examples.

Now, I understand both points. Having them non placeable is a bad idea because players will not only want so see the items but potentially also try out the blocks in the world. And if they have custom nbt and we scan them in, this would be a pain. But, inevitably people will also wonder about the other blocks.

Why can't we leave the random blocks as is and add a custom creative tab with a short description? "DO adds the following block types:**** with numerous variants of each. above you see a randomly generated selection. Use the stone cutter with different input materials to generate them"

That both makes clear they are not the only ones, let's the player use them + is excellent documentation

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by marchermans** [comment]
> The JEI solution is fine and displaying random example blocks there is fine.

That would likely be a good direction.
However what about this variation: Those special stacks, are literally random. They place random variations? 
Would be a proper mechanic to show that they are special. And when they are scanned they are not special anymore, since when they are placed they get a fixed variant?

**Comment by marchermans** [comment]
> > The JEI solution is fine and displaying random example blocks there is fine.
> 
> That would likely be a good direction. However what about this variation: Those special stacks, are literally random. They place random variations? Would be a proper mechanic to show that they are special. And when they are scanned they are not special anymore, since when they are placed they get a fixed variant?

Besides this, why can we not just dynamically inject the stacks into JEI, instead of showing them in the creative menu?
I mean the point of these stacks is to directly allow for them to be used to find the relevant recipes is it not?. Then why are we including them in the creative menu in the first place?

**Comment by Raycoms** [comment]
Okay, we discussed this further internally: TL;DR: All on JEI side is fine, nothing to change there. On the creative tab front, for this PR to be merged, we should have a special creative tab with a short explanation on the mod where this explanation also points out that the blocks in the tab are merely random example blocks and that all additional variants can be created through the cutter block.

**Comment by uecasm** [comment]
> Besides this, why can we not just dynamically inject the stacks into JEI, instead of showing them in the creative menu?

Because, as I already said, JEI will only show stacks that came from the creative tab.  That is its rule for deciding whether a mod or modpack wants to hide items or not.  This is the main reason why I added them in the first place, because otherwise they don't show up in JEI -- or more specifically, you can do usage searches, or look up a recipe if you already have one, but it's impossible to look up recipes when you don't have one, which is the most important use case.  (I don't think it's bad that they also show up in the creative tabs, but I almost never use those.)

> On the creative tab front, for this PR to be merged, we should have a special creative tab with a short explanation on the mod where this explanation also points out that the blocks in the tab are merely random example blocks and that all additional variants can be created through the cutter block.

I don't think any extra explanation is useful.  Every item in the creative tab already shows a full tooltip explaining that it is created in an architect's cutter and the precise ingredients required to make that particular item.  The only slightly surprising thing is that each session you will see different blocks in the list, and that's something that is already implied by the fact that each block already uses wildly different ingredients, and it is self-evident after two or three sessions that they're being randomly generated each time.  And this is also obvious if you look up the recipe of any of the items in JEI, because it cycles between all the different inputs.

**Comment by uecasm** [comment]
Having said that, I could put in an information page for the architect's cutter (this appears on an info tab when you look up the recipe for the cutter itself, which also happens if you click on the left-side tab showing the cutter when looking at any of its recipes).

(See the compost barrel in MineColonies for an example of such a page.)

This could explain the basic idea of how the cutter works, similar to the above.  I don't mind adding something like that in, though as I said I don't think it's really needed -- once you've actually seen a few recipes and played with it a little it's pretty obvious how it all behaves.

**Comment by Raycoms** [comment]
> I don't think any extra explanation is useful. Every item in the creative tab already shows a full tooltip explaining that it is created in an architect's cutter and the precise ingredients required to make that particular item. The only slightly surprising thing is that each session you will see different blocks in the list, and that's something that is already implied by the fact that each block already uses wildly different ingredients, and it is self-evident after two or three sessions that they're being randomly generated each time. And this is also obvious if you look up the recipe of any of the items in JEI, because it cycles between all the different inputs.

The idea is that this explanation tells you specifically that there are millions of variants each and that you can craft ton of other variants. A lot of creative design players will use the mod, and I believe not all of these people use JEI (as they don't need recipes). So a quick and short explanation here would be good.

**Comment by uecasm** [comment]
> The idea is that this explanation tells you specifically that there are millions of variants each and that you can craft ton of other variants. A lot of creative design players will use the mod, and I believe not all of these people use JEI (as they don't need recipes). So a quick and short explanation here would be good.

How, though?  There is no space to do so on the creative tab itself.  It does not have any custom rendering hooks for the page itself, other than setting a background texture (which is not suitable for text display).  It is possible to add additional tooltips to the items, but the existing tooltips are already quite large (especially at default gui scale), already contain most of what you need to know anyway, and doing that requires adding the code to each and every item type individually (which is doable, but annoying, because they don't have a common base).  And if it's done by putting special NBT on the items instead of via more sneaky means, that will also persist into items pulled out of the creative tabs and will be entirely counterproductive.  (Unless as Orion suggested, this pulls out a rainbow block that places random textures, but that seems even more counterproductive.)

Anyone downloading the mod on its own should already be well aware that it adds millions of block variants, since it's right there in the description.  Anyone getting it as part of a modpack may be less aware, but that's what the wiki is for.  And modpacks will always contain JEI.

All in all, as I said before, it's a lot of effort to put in for something which is utterly useless once people get over the initial surprise -- which won't take very long, if they're surprised at all.  A lot of mods do this sort of thing.

**Comment by Raycoms** [comment]
> > The idea is that this explanation tells you specifically that there are millions of variants each and that you can craft ton of other variants. A lot of creative design players will use the mod, and I believe not all of these people use JEI (as they don't need recipes). So a quick and short explanation here would be good.
> 
> How, though? There is no space to do so on the creative tab itself. It does not have any custom rendering hooks for the page itself, other than setting a background texture (which is not suitable for text display). It is possible to add additional tooltips to the items, but the existing tooltips are already quite large (especially at default gui scale), already contain most of what you need to know anyway, and doing that requires adding the code to each and every item type individually (which is doable, but annoying, because they don't have a common base). And if it's done by putting special NBT on the items instead of via more sneaky means, that will also persist into items pulled out of the creative tabs and will be entirely counterproductive. (Unless as Orion suggested, this pulls out a rainbow block that places random textures, but that seems even more counterproductive.)
> 
> Anyone downloading the mod on its own should already be well aware that it adds millions of block variants, since it's right there in the description. Anyone getting it as part of a modpack may be less aware, but that's what the wiki is for. And modpacks will always contain JEI.
> 
> All in all, as I said before, it's a lot of effort to put in for something which is utterly useless once people get over the initial surprise -- which won't take very long, if they're surprised at all. A lot of mods do this sort of thing.

I thought there was an easy way to alter the creative tab rendering by adding extra text, background etc.

**Comment by uecasm** [comment]
> I thought there was an easy way to alter the creative tab rendering by adding extra text, background etc.

Background, yes, but that doesn't alter the slot positions so you can't do much more than change the colour or have fancy decorative bits.

For the rest, as far as I can tell, no.  Although you're welcome to prove me wrong. 😸

**Comment by marchermans** [comment]
> > Besides this, why can we not just dynamically inject the stacks into JEI, instead of showing them in the creative menu?
> 
> Because, as I already said, JEI will only show stacks that came from the creative tab. That is its rule for deciding whether a mod or modpack wants to hide items or not. This is the main reason why I added them in the first place, because otherwise they don't show up in JEI -- or more specifically, you can do usage searches, or look up a recipe if you already have one, but it's impossible to look up recipes when you don't have one, which is the most important use case. (I don't think it's bad that they also show up in the creative tabs, but I almost never use those.)
> 
> > On the creative tab front, for this PR to be merged, we should have a special creative tab with a short explanation on the mod where this explanation also points out that the blocks in the tab are merely random example blocks and that all additional variants can be created through the cutter block.
> 
> I don't think any extra explanation is useful. Every item in the creative tab already shows a full tooltip explaining that it is created in an architect's cutter and the precise ingredients required to make that particular item. The only slightly surprising thing is that each session you will see different blocks in the list, and that's something that is already implied by the fact that each block already uses wildly different ingredients, and it is self-evident after two or three sessions that they're being randomly generated each time. And this is also obvious if you look up the recipe of any of the items in JEI, because it cycles between all the different inputs.

What you are saying here is just not true on all fronts.
First and foremost, you do not need to add them to the creative tab to get them in JEI.
<https://github.com/ChiselsAndBits/Chisels-and-Bits/blob/55b5f241cc989bbcaf03644dc4359112b7fafb15/src/main/java/mod/chiselsandbits/integration/jei/JEIPlugin.java#L34-L48> I have been doing it for years this way, when I want dynamic injection of items.

Secondly, and much more importantly, I have been around the block before in the minecraft community. I have been modding for over 10 years, and in my opinion players looking for blocks, which are not clearly labelled as to where to find them, when they can find others, will lead to a massive amount of complaining. This complaining is what I explicitly do not want again, because I had to deal with this one way or another on every of our mods, especially since we discussed this topic now twice internally and now clearly defined as to how we want this implemented.

Thirdly to address other concerns in other posts here.
You can use the events forge fires to render the text in the UI if you are so inclined!
Yes this is more work, but this is what we want. If you are not willing to spend the time making it in a way that everybody is happy with your contribution, then I am sorry but the contribution is not worth our time. Since you are then settling us and our community management team up with the burden of dealing with your lack of commitment to the PR and code you created.

This said, I would like a final answer on this from you, a simple yes or no: Are you going to address our concerns with regards to the issue of the items in that creative tab or not, because If not I will do it myself eventually, or we will assign this ourselfs to somebody else, who understands that sometimes tasks that we have todo are not perfect and that these will need to be handled with care.

**Comment by Raycoms** [comment]
Basically: Custom background that disables half of the slots, cusotm render event to render some text over the empty space in the custom background. This should be very straightforward, and the text can easily be localized and adjusted to needs.

**Comment by uecasm** [comment]
There's always going to be players who don't read.  You just have to ignore those, or point them at JEI.  Which is easier when JEI integration actually exists.

But fine, if you're not interested in improving usability of your mod, then I don't care any more.  Do whatever you like.  I'm not going to add pointless code that I don't want to support either.

**Pull request closed without merging**
